src/server.c

```diff
@@ -4124,6 +4125,14 @@ int processCommand(client *c) {
         }
     }
 
+    /* Check if the command has cross slot keys, incompatible with cluster mode. */
+    if (server.cluster_compatibility_sample_ratio && !server.cluster_enabled &&
+        !(!(c->cmd->flags&CMD_MOVABLE_KEYS) && c->cmd->key_specs_num == 0) &&
```

Comment:
shouldn't we use `doesCommandHaveKeys`?
or maybe this one also applies for sharded pubsub channels?

Reply:
i saw we use this method to check slot node
https://github.com/redis/redis/blob/b704179f158feef916cca9ba1907a1d0e817886c/src/server.c#L4108
`doesCommandHaveKeys` seems better

Reply:
Please check implications for pubsub too

Reply:
`doesCommandHaveKeys` doesn't consider shard pubsub channel commands have keys, since they have key specifications and marked as NOT_KEY, such as `ssubscribe` command
https://github.com/redis/redis/blob/unstable/src/commands/ssubscribe.json#L22
but i think here we should check shard pubsub channel commands, it is not compatible with cluster mode if all the specified shard channels don't belong to a single slot

Reply:
i will add test cases for shard pubsub channel commands later